src/libraries/System.Linq/src/System/Linq/MaxMin.cs

```diff
@@ -90,6 +91,30 @@ private interface IMinMaxCalc<T> where T : struct, IBinaryInteger<T>
                         }
                     }
                 }
+                else
+                {
+                    ref T current = ref MemoryMarshal.GetReference(span);
+                    ref T lastVectorStart = ref Unsafe.Add(ref current, span.Length - Vector512<T>.Count);
+
+                    Vector512<T> best = Vector512.LoadUnsafe(ref current);
+                    current = ref Unsafe.Add(ref current, Vector512<T>.Count);
+
+                    while (Unsafe.IsAddressLessThan(ref current, ref lastVectorStart))
+                    {
+                        best = TMinMax.Compare(best, Vector512.LoadUnsafe(ref current));
+                        current = ref Unsafe.Add(ref current, Vector512<T>.Count);
+                    }
+                    best = TMinMax.Compare(best, Vector512.LoadUnsafe(ref lastVectorStart));
+
+                    value = best[0];
+                    for (int i = 1; i < Vector512<T>.Count; i++)
+                    {
+                        if (TMinMax.Compare(best[i], value))
+                        {
+                            value = best[i];
+                        }
+                    }
```

Comment:
Not for this PR, but I think we should revisit the conditions when the vectorized code-paths are entered and how the reduction to the scalar value is performed.

For e.g. `int` it's 16 iterations of that loop, and when (by accident) the input span is 16 ints long, then the whole vectorized min/max is pointless.

Maybe special case some common types like `int`, where (at least with Sse4.1) a vectorized Min version exists, so the Vec512 could be split into 4x Vec128 and continue vectorized (min + shuffling around). Or isn't that it worth for Linq and / or not worth the trouble for the specialization for some types?

Reply:
I agree.. Probably the vectorized code path should only be entered if the number of elements is two or 4 time the vector length. But the additional checks would introduce additional overhead as well.

Regarding the Min/Max over the vector in the last code part: This could probably be optimized as well via vectorization for the `Vector512` and `Vector256` cases, by loading the Vector512 `best` Vector into two Vector256 doing the Min/Max there via SIMD and same with Vector256 -> Vector128..

Not sure how large the overhead on the actual target platforms is to do this, as we would switch between AVX512 to AVX256 (probably cheap) and then to SSE 4.2... If i remember correctly switching from SSE to AVX is quite expensive on some older Intel platforms.

Reply:
> switching from SSE to AVX is quite expensive on some older Intel platforms

As you said "on older" ones, and `vzeroupper` can help here.

Maybe it's best to just keep the code as is (in your PR) now, and later overhaul the codebase once `ISimdVector` et. al. has proper helpers for reduction, etc. I remember there was a discussion in one issue / PR about such things, but can't find it ATM.